File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxi = INT_MIN;
    
    int findMaxPathSum(TreeNode* root) {
      if (root == NULL) return 0;

      int leftMaxPath = max(0, findMaxPathSum(root -> left));
      int rightMaxPath = max(0, findMaxPathSum(root -> right));
      int val = root -> val;
      maxi = max(maxi, (leftMaxPath + rightMaxPath) + val);
      return max(leftMaxPath, rightMaxPath) + val;

    }

    int maxPathSum(TreeNode* root) {
        findMaxPathSum(root);

        return maxi; 
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.ans = float('-inf')

        def dfs(root, curSum):
            if not root:
                return 0
            
            # check with just root
            self.ans = max(self.ans,root.val)

            leftSum = dfs(root.left, curSum)
            rightSum = dfs(root.right, curSum)

            # check with root + left
            self.ans = max(self.ans, root.val + leftSum)
            
            # check with root + right
            self.ans = max(self.ans, root.val + rightSum)
            
            # check with root + left + right
            self.ans = max(self.ans, root.val + rightSum + leftSum)
            
            #return max(root,root + left,root + right)
            return max(root.val,root.val + leftSum, root.val + rightSum)
        

        dfs(root,0)

        return self.ans